From 365c7cc34874f5d6fdbf66073a00b5f837d53f62 Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Fri, 3 Oct 2014 21:26:00 -0400
Subject: [kernel] printk: Add function to return log buffer address and size

Message-id: <20141003212600.20407.6998.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 96895
O-Subject: [PATCH RHEL7.1 BZ1142244 1/4] printk: Add function to return log buffer address and size
Bugzilla: 1142244
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1142244

Description:
------------
Platforms like IBM Power Systems supports service processor
assisted dump. It provides interface to add memory region to
be captured when system is crashed.

During initialization/running we can add kernel memory region
to be collected.

Presently we don't have a way to get the log buffer base address
and size. This patch adds support to return log buffer address
and size.

Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Acked-by: Andrew Morton <akpm@linux-foundation.org>

RHEL Version Found:
-------------------
7.0

Upstream Status:
----------------
https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=14c4000a88afaaa2d0877cc86d42a74fde0f35e0

----------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/printk.h b/include/linux/printk.h
index 22c7052..2deb0b0 100644
--- a/include/linux/printk.h
+++ b/include/linux/printk.h
@@ -9,6 +9,9 @@
 extern const char linux_banner[];
 extern const char linux_proc_banner[];
 
+extern char *log_buf_addr_get(void);
+extern u32 log_buf_len_get(void);
+
 static inline int printk_get_level(const char *buffer)
 {
  if (buffer[0] == KERN_SOH_ASCII && buffer[1]) {
diff --git a/kernel/printk.c b/kernel/printk.c
index 69b0890..81ef6e3 100644
--- a/kernel/printk.c
+++ b/kernel/printk.c
@@ -258,6 +258,18 @@ static u32 log_buf_len = __LOG_BUF_LEN;
 /* cpu currently holding logbuf_lock */
 static volatile unsigned int logbuf_cpu = UINT_MAX;
 
+/* Return log buffer address */
+char *log_buf_addr_get(void)
+{
+ return log_buf;
+}
+
+/* Return log buffer size */
+u32 log_buf_len_get(void)
+{
+ return log_buf_len;
+}
+
 /* human readable text of the record */
 static char *log_text(const struct log *msg)
 {
-- 
1.7.1